### DISCUSSION TITLE
Is it possible to save tflint output in a file? Might be useful when used in a pre-commit hook

### DISCUSSION THREAD

**Discussion by JuanJoanJon** [terraform-linters/tflint#1581] [Q&A] [answered]
Is it possible to save tflint output in a file? Might be useful when used in a pre-commit hook.

**Comment by bendrucker** [comment, marked as answer]
Not offered. Is there a reason you wouldn't just use a shell?

```
tflint > results
```

Not aware of any linter tools that offer file output as described, both because it's an uncommon need and it's readily handled via shell.

**Reply by JuanJoanJon** [reply]
> Not offered. Is there a reason you wouldn't just use a shell?

When using tflint in a CI pipeline I'd like to write the output (a report) in a specific file (and if this is configured by tflint command args, the better). This report will be then consumed by another tool during CI pipeline execution.

> Not aware of any linter tools that offer file output as described, both because it's an uncommon need and it's readily handled via shell.

Checkov (not exactly a linter but kind of) can do that with `--output-file-path` command arg

**Reply by bendrucker** [reply]
Thanks, but this is just a rephrasing of the original request

> and if this is configured by tflint command args, the better

Why would you want this? What is gained by passing a flag versus using a shell?

I looked at the Checkov example (links would be appreciated) and the flag appears to be an escape hatch introduced by other complexity that doesn't currently apply to TFLint:

https://www.checkov.io/2.Basics/CLI%20Command%20Reference.html

> Name of the output folder to save the chosen output formats. Advanced usage: By using -o cli -o junitxml –output-file-path console,results.xml the CLI output will be printed to the console and the JunitXML output to the file results.xml.

In other words, Checkov is offering to print user-friendly output to stdout while writing machine output to a file. _This_ is an answer to my question about why you'd want this.

That said, I don't see TFLint pursuing the broader feature request at the moment, for multiple output formats directed to different locations. And if we did pursue it, I don't particularly care for Checkov's interface, it might look something more like this:

```
tflint --file-output ./foo --file-output-format json
```

If `--format` and `--file-output` were set and `--file-output-format` were unset, `--format` would be used for both stdout and the file.

**Reply by JuanJoanJon** [reply]
I appreciate your time for trying to understand the root of my request. Actually you explained it far better than me.

To complement the original request, this came to me while using [pre-commit-terraform](https://github.com/antonbabenko/pre-commit-terraform), being [terraform_tflint](https://github.com/antonbabenko/pre-commit-terraform#terraform_tflint) one of hooked tools which doesn't allow pretty-printing/machine-printing to a file. Given that many of the other tools ([tfsec](https://github.com/aquasecurity/tfsec), [terrascan](https://github.com/tenable/terrascan), [checkov](https://github.com/bridgecrewio/checkov), [infracost](https://github.com/infracost/infracost)) allow that feature it looked like a "standalone" capability, specially when run as pre-commit hook where shell redirections are difficult to achieve.

I do agree with your interface proposal. May I kindlly ask you the reason why you don't see pursuing the broader feature at the moment?

**Reply by bendrucker** [reply]
Thanks for understanding. There's a big backlog of potential work to do across both core and rulesets and niche requests like this are unlikely to end up high on the list anytime soon. Shell is a viable option for integrations that want to redirect the output to disk. And even for multiple formats, you can just run multiple times, once per desired format. Wastes a bit of time but unless your modules are very large it may not be meaningful. Could even do the runs concurrently. 

If you want to iterate further on the design and author a PR to implement it, I'll make time to review.
